torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -505,9 +505,62 @@ Tensor masked_fill_backward(const Tensor& grad, const Tensor& mask) {
       : grad.masked_select(mask).sum();
 }
 
-Tensor mul_tensor_backward(Tensor grad, Tensor other, ScalarType self_st) {
-  auto out = grad * other.conj();
-  return handle_r_to_c(self_st, out);
+Tensor mul_tensor_backward(
+    const Tensor& grad,
+    const Tensor& other,
+    const Tensor& self,
+    ScalarType self_st) {
+  // If a.is_sparse and b.is_sparse, then mul(a, b) expects:
+  // 1. a.dim() == b.dim(),
+  // 2. a.shape[:a.sparse_dim] == b.shape[:b.sparse_dim].
+  // Note, however, that mul(a, b) will handle broadcasting in dense dims.
+  // Autograd, however, will have issues with that and will try running
+  // either a.sum(d=...) or b.sum(d=...) to propagate sparse grads of the
+  // same shape as the inputs'. However, sum(d=...) is not implemented
+  // for sparse tensors. So, instead, we explicitly reduce over grads'
+  // dense dims and create new sparse gradient tensors that now match
+  // the shape of the inputs.
+  const auto handle_sparse_sparse_case = [](const Tensor& self_grad,
+                                            const Tensor& self) -> Tensor {
+    // No broadcasting in dense dims, no need to modify anything.
+    if (self_grad.sizes() == self.sizes()) {
+      return self_grad;
+    }
+    // Here self dense dims broadcast over self_grad dense dims.
+    // This means we need to reduce over broadcasted dims in self_grad.values
+    // such that self_grad.values.sum(...).shape == self.values.shape,
+    // otherwise autograd would try running self_grad.sum(dim=...), and
+    // sum(dim=...) is not implemented for sparse tensors.
+    const auto self_sparse_dim = self.sparse_dim();
+    const auto self_grad_indices = self_grad.indices().clone();
```

Comment:
This might be the source of some of the CI failures? Do we have any guarantees that self_grad will always be coalesced?

Reply:
I thought I saw some errors complaining about accessing indices on uncoalesced? 

![image](https://user-images.githubusercontent.com/9041474/191324745-56b4ba43-f653-463c-9e19-d4b4008f7a97.png)

Reply:
This is weird. Is this a bug or a feature? CC @cpuhrsch , @pearu 
```python
In [1]: import torch

In [2]: idx = torch.tensor([[1, 1, 1, 2, 2, 2], [2, 2, 1, 3, 3, 3]], dtype=torch.long)

In [3]: val = torch.tensor([True, False, True, False, False, True])

In [4]: s = torch.sparse_coo_tensor(idx, val, size=(10, 10))

In [5]: s.indices()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 s.indices()

RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first

In [6]: s._indices()
Out[6]: 
tensor([[1, 1, 1, 2, 2, 2],
        [2, 2, 1, 3, 3, 3]])

```

Reply:
I have replaced `.indices()` with `._indices()` to resolve that.